Entry-level tutorial on binary search (#1411)
informalsystems/apalache: docs/src/tutorials/entry-tutorial.md

# Entry-level Tutorial on the Model Checker

> **Difficulty: Blue trail -- Easy**

Reviewer: ```suggestion
**Difficulty: Blue trail – Easy**
```

`mdbook` doesn't do latex-style ligatures, use UTF en dash

In this tutorial, we show how to turn an implementation of binary search into a
TLA+ specification. This implementation is known to have an out-of-bounds
error, which once existed in Java, see [Nearly All Binary Searches and

> Mergesorts are Broken][] by Joshua Bloch (2006). Our goal is not to write a

Reviewer 2: ```suggestion
Mergesorts are Broken][] by Joshua Bloch (2006). Our goal is to write a
```

Author: Good point. I've rephrased it.

specification of an abstract binary search algorithm. You can find such a
specification and a proof in [Proving Safety Properties][] and [Binary search
with a TLAPS proof][] by [Leslie Lamport][] (2019).

This tutorial is written under the assumption that the reader does not have any
knowledge of TLA+ and Apalache. Since we are not diving into protocol and
algorithm specifications too quickly, this is a nice example to start with.

> Sometimes, we refer to the model checker TLC in this text. TLC is the standard
> model checker for TLC, which was introduced in the late 90es.  If you are new

Reviewer: ```suggestion
Sometimes, we refer to the model checker TLC in this text. TLC is another
model checker for TLA+ and was introduced in the late 90s.  If you are new
```

to TLA+ and want to learn more about TLC, check the [TLC][] project and the

> [TLA+ Video Course][] by Leslie Lamport.  If you are an experienced TLC user,

Reviewer 2: ```suggestion
[TLA+ Video Course][] by Leslie Lamport. If you are an experienced TLC user,
```

you will find this tutorial helpful too, as it demonstrates the strong points
of Apalache.

We discover that the same bug not only causes the out-of-bounds error, but also
prevents the algorithm from terminating in `log2(high - low + 1)`, as it is

> usually expected.

Reviewer: For me, this paragraph feels out of context here. You may want to merge it with the first paragraph above, or just move it to the relevant section below.

Author: True. I have expanded this paragraph and merged it with the second paragraph.

## Related documents

>  - [Tutorial on Snowcat][] shows how to write type annotations.

Reviewer: ```suggestion
 - [Tutorial on Snowcat][] shows how to write type annotations for Apalache.
```

>  - [TLA+ Cheatsheet in HTML][] summarizes the common TLA+ constructs.  If you
>    prefer the printable version in pdf, check the [Summary of TLA+][].

Reviewer: ```suggestion
 - [TLA+ Cheatsheet in HTML][] summarizes the common TLA+ constructs.  If you
   prefer a printable version in pdf, check the [Summary of TLA+][].
```

## Setup

We assume that you have Apalache installed. If not, check the manual page on
[Apalache installation][]. The minimal required version is 0.22.0.

## Running example: Binary search

We are not going to explain the idea of binary search in this tutorial. If you
need more context on this, check the Wikipedia page on the [Binary search

> algorithm][]. Let's jump straight in the Java code that is given in [Nearly All

Reviewer: ```suggestion
algorithm][]. Let's jump straight into the Java code that is given in [Nearly All
```

Binary Searches and Mergesorts are Broken][]:

```java
1:     public static int binarySearch(int[] a, int key) {
2:         int low = 0;
3:         int high = a.length - 1;
4:
5:         while (low <= high) {
6:             int mid = (low + high) / 2;
7:             int midVal = a[mid];
8:
9:             if (midVal < key)
10:                 low = mid + 1
11:             else if (midVal > key)
12:                 high = mid - 1;
13:             else
14:                 return mid; // key found
15:         }
16:         return -(low + 1);  // key not found.
17:     }
```

> Fast forward, as was found by Joshua Bloch, the addition in line 6 may throw

Reviewer 3: I'm not sure what "fast forward" means here?

Author: Removed.

> Fast forward, as was found by Joshua Bloch, the expression in line 6 may throw
> an out of bounds exception, due to an integer overflow. This is because `low`

Reviewer: ```suggestion
Fast forward, as was found by Joshua Bloch, the addition in line 6 may throw
an out of bounds exception at line 7, due to an integer overflow. This is because `low`
```

and `high` are signed integers whose maximum value could not grow above `2^31 -
1`, so the sum `low + high` could wrap into a negative number.

Let's see how TLA+ and Apalache can help us here. A bit of warning: The final
TLA+ specification will happen to be longer than the 17 lines above. Don't get
disappointed too fast. There are several reasons for that:

 1. TLA+ is not tuned towards one particular class of algorithms, e.g.,
 sequential algorithms.

 2. Related to the previous point, TLA+ and Apalache are not tuned to C or Java
 programs. A software model checker such as [CBMC][], [Stainless][], or
 [Coral][] would probably accept a shorter program, and it would check it
 faster. However, if you have a sledgehammer like TLA+, you don't have to learn
 other languages.

 3. We have to explicitly write the expected properties that are normally

>  written in comments.

Reviewer: I would rephrase this point into something like "We will explicitly state the expected properties of the algorithm to be checked by Apalache. In imperative languages, these properties are usually omitted or written as plain-text comments."

 4. We have to introduce a bit of boilerplate, to make Apalache work.

## Step 0: Introducing a template module

> Let's open a new file called `BinSearch0.tla` and type the very minimal module

Reviewer: ```suggestion
Let's open a new file called `BinSearch0.tla` and type a very minimal module
```

definition:

```tla
{{#include ../../../test/tla/bin-search/BinSearch0.tla:1:8}}
```

This module is doing almost nothing. However, it contains a few important things:

>  - It uses three standard modules: `Integers`, `Sequences`, and `Apalache`.

Reviewer: I would either `s/uses/imports/` or provide a one-sentence explanation.

 - It declares the predicate `Init`. This predicate describes the initial

>    states of our state machine. Since we have not declared any variables, it

Reviewer: So far, you haven't touched on the idea that we're going to encode a state machine. Might be good to explain in a sentence in the introductory paragraph about TLA+.

Reviewer 2: I also think that is important to do

Author: Added a short explanation.

>    defines the only single state.

Reviewer: ```suggestion
   defines the single possible state.
```

 - It declares the predicate `Next`. This predicate describes the transitions
   of our state machine. Again, there are no variables and `Next == TRUE`,

>    so this transition goes from the initial state to itself.

Reviewer: I mean, technically this isn't true for pure TLA+, right? It goes to any unconstrained state.

Author: Right. But since there are no variables, there should be just one state.

Now it is a good time to check that Apalache is working. Run the following command:   

```sh
$ apalache-mc check BinSearch0.tla
```

The tool output is a bit verbose. Below, you can see the important lines of the
output:

```
...
PASS #13: BoundedChecker
Step 0: picking a transition out of 1 transition(s)
Step 1: picking a transition out of 1 transition(s)
...
Step 10: picking a transition out of 1 transition(s)
The outcome is: NoError
Checker reports no error up to computation length 10
... 
```

> We can see that Apalache is running and it finds no error. Not unexpected.

Reviewer: ```suggestion
We can see that Apalache is running and it finds no error, as expected.
```

If you are curious, replace `TRUE` with `FALSE` in either `Init` or `Next`,
run Apalache again and observe what is happening.

It is usually a good idea to start with a spec like `BinSearch0.tla`, to ensure
that the tools are working.

TODO: add the link to BinSearch0.tla when we have it on github.

## Step 1: Introducing specification parameters

> Copy `BinSearch0.tla` to `BinSearch1.tla` and open `BinSearch1.tla`.

Author: It's a tutorial. You can do it without looking into sources.

```

The benefit of defining the bit width as a parameter is that we can try
our specification for various bit widths of integers: 4-bit, 8-bit, 16-bit, 32-bit,
etc.  Similar to many programming languages, we introduce several constant
definitions (`a^b` is `a` taken to the power of `b`):

```tla
{{#include ../../../test/tla/bin-search/BinSearch1.tla:24:30}}
```

Note that these definitions do not constrain integers in any way. They are
simply convenient names for the constants that we will need in the
specification.

To make sure that the new specification does not contain syntax errors or
type errors, execute:

```sh
$ apalache-mc check BinSearch1.tla
```

TODO: add the link to BinSearch1.tla when we have it on github.

## Step 2: Specifying the base case

Copy `BinSearch1.tla` to `BinSearch2.tla` and open `BinSearch2.tla`.

We start with the simplest possible case that occurs in `binarySearch`. Namely,
we consider the case of `low > high`, that is, `binarySearch` is never entering
the loop.

**Introduce variables**. To do that, we have to finally introduce some
variables. Obviously, we have to introduce variables `low` and `high`. This is
how we do it:

```tla

> {{#include ../../../test/tla/bin-search/BinSearch2.tla:33:38}}

Reviewer: ```suggestion
{{#include ../../../test/tla/bin-search/BinSearch2.tla:32:38}}
```

Start with the `VARIABLES` block

```

The variables `low` and `high` are called *state variables*. They are always
present in a state of our state machine. That is, they are never introduced and
never removed. Remember, TLA+ is not tuned towards any particular computer
architecture and thus it does not even have the notion of an execution stack.
You can think of `low` and `high` as being global variables. Yes, global
variables are generally a bad smell in a program. However, when dealing with a
specification, they are much easier to reason about than the execution stack.
If you need to introduce local definitions, you will see how to do that later
in this tutorial.

Less obviously, we introduce two additional variables:

```tla
{{#include ../../../test/tla/bin-search/BinSearch2.tla:39:44}}
```

> The variable `isTerminated` indicates, whether our search has terminated. Why
> did we even have to introduce it? Because, in general, computer systems do not

Reviewer: ```suggestion
The variable `isTerminated` indicates whether our search has terminated. Why
do we even have to introduce it? Because, in general, computer systems do not
```

have to terminate. If we want to specify the Internet or Bitcoin, do we

> understand what it means for them to terminate?

Reviewer: I don't understand the last sentence.

Author: Right. I wanted to say that some distributed systems are not designed to terminate. Maybe I should just write that :)

> The variable `returnValue` will contain the result of the binary search, when
> the search terminated. Recall, there is no execution stack. Hence, we introduce

Reviewer: ```suggestion
The variable `returnValue` will contain the result of the binary search, when
the search terminates. Recall, there is no execution stack. Hence, we introduce
```

the variable `returnValue` right away. The downside is that we have to do
book-keeping for this variable.

**Initialize variables.** Having introduced the variables, we have to
initialize them. That is, we want to specify lines 2-3 of the Java code:

```java
1:     public static int binarySearch(int[] a, int key) {
2:         int low = 0;
3:         int high = a.length - 1;
           ...
17:    }
```

To this end, we change the body of the predicate `Init` to the following:

```tla
Init ==
  low = 0 /\ high = Len(INPUT_SEQ) - 1 /\ isTerminated = FALSE /\ returnValue = 0
```

> You probably have guessed, what the above line means. You are probably a bit
> puzzled about the mountain-like operator `/\`. It is called *conjunction*,

Reviewer: ```suggestion
You probably have guessed, what the above line means. You might be a bit
puzzled about the mountain-like operator `/\`. It is called *conjunction*,
```

> puzzled about the mountain-like operator `/\`. It is called *conjunction*,

Reviewer: We could rewrite this to `Maybe you're a bit puzzled about`

which is usually written as `&&` or `and` in programming languages. The
important effect of the above expression is that every variable in the
left-hand side of `=` is required to have the value specified in the right-hand
side of `=`. It is important to know that TLA+ does not impose any particular
order of evaluation for `/\`. However, both Apalache and TLC evaluate
some expressions of the form `x = e` in the initialization predicate as
assignments. Hence, it is often a good idea to think about `/\` as being

> evaluated from left to right.

Reviewer: This might be too technical at this point. Or maybe we could make it into a sidenote in a box or something?

As it is hard to fit many expressions in one line, TLA+ offers special syntax
for writing a big conjunction. Here is the standard way of writing `Init`
(indentation is important):

```tla
{{#include ../../../test/tla/bin-search/BinSearch2.tla:46:51}}
```

The above lines do not deserve a lot of explanation. As you have guessed,
`Len(INPUT_SEQ)` computes the length of the input sequence.

**Update variables.** Having done all the preparatory work, we are now ready to
specify the behavior in lines 5 and 16 of `binarySearch`.

```java
1:     public static int binarySearch(int[] a, int key) {
2:         int low = 0;
3:         int high = a.length - 1;
4:
5:         while (low <= high) {
            ...
15:        }
16:        return -(low + 1);  // key not found.
17:    }
```

To this end, we redefine `Next` as follows:

```tla
{{#include ../../../test/tla/bin-search/BinSearch2.tla:53:62}}
```

Most likely, you have no problem reading this definition, except for the part
that includes `isTerminated'`, `returnValue'`, and `UNCHANGED`. The expression

> `isTerminated' = TRUE` means that `isTerminated` has the value `TRUE` in the
> next state.  Likewise, `returnValue' = -(low + 1)` means that `returnValue` has

Reviewer: ```suggestion
`isTerminated' = TRUE` means that `isTerminated` shall have the value `TRUE` in the
next state of our state machine.  Likewise, `returnValue' = -(low + 1)` means that `returnValue` has
```

> next state of our state machine.  Likewise, `returnValue' = -(low + 1)` means that `returnValue` has

Author: Wrote a slightly fixed version

the value `-(low + 1)` in the next state. The expression `UNCHANGED <<low,

> high>>` is a convenient shortcut for writing `low' = low /\ high' = high`.

Reviewer: You may want to expand on why having `UNCHANGED` is important. It will not be obvious for the level of reader targeted here.

It is important to understand that an expression like `returnValue' = -(low +

> 1)` *does not immediately update* the variable in the left-hand side. Hence,

Reviewer: ```suggestion
1)` *does not immediately update* the variable on the left-hand side. Hence,
```

being delayed until the whole predicate `Next` is evaluated.

TODO: add the link to BinSearch2.tla when we have it on github.

## Step 3: Basic checks for the base case

As we discussed, it is a good habit to run the model checker as you are writing
the specification. Even if it would not check much, you would be able to catch
the moment when the model checker would slow down. This may give you a useful
hint about changing a few things before you have written too much code.

Let us check `BinSearch2.tla`:

```sh
$ apalache-mc check BinSearch2.tla
```

If it is your first TLA+ specification, you may be surprised by this error:

```
...
PASS #13: BoundedChecker
This error may show up when CONSTANTS are not initialized.
Check the manual: https://apalache.informal.systems/docs/apalache/parameters.html
Input error (see the manual): SubstRule: Variable INPUT_SEQ is not assigned a value
...
```

Apalache complains that we have defined several parameters (`INPUT_SEQ`,
`INPUT_KEY`, and `INT_WIDTH`) but we have never defined them.

**Adding a model file.** The standard approach in this case is either to fix
all constants in a TLC configuration file, or to introduce another module that

> fixes the parameters and instantiates the general specification.  Although

Reviewer 2: ```suggestion
fixes the parameters and instantiates the general specification. Although
```

told us that it has not found any contradictions. But it did not tell us
anything interesting about our expectations. Because we have not set our
expectations yet!

TODO: add the link to BinSearch2.tla and MC2_8.tla when we have them on github.

## Step 3b: Specifying an invariant and checking it for the base case

Copy `MC2_8.tla` to `MC3_8.tla`. Copy `BinSearch2.tla` to `BinSearch3.tla` and
open `BinSearch3.tla`.

What do we expect from the binary search? We can check the Java documentation,
e.g., [Arrays.java in OpenJDK][]:

> ...the return value will be >= 0 if and only if the key is found.

This property is actually quite easy to write in TLA+. First, we
introduce the property that we call `ReturnValueIsCorrect`:

```tla
{{#include ../../../test/tla/bin-search/BinSearch3.tla:72:83}}
```

Let us decompose this property into smaller pieces. First, we define the set
`MatchingIndices`:

```tla
{{#include ../../../test/tla/bin-search/BinSearch3.tla:75:77}}
```

With this TLA+ expression we define a *local constant* called `MatchingIndices`
that is equal to the set of indices `i` in `INPUT_SEQ` so that the sequence

> elements defined with these indices are equal to `INPUT_KEY`.  If this syntax

Reviewer 2: ```suggestion
elements defined with these indices are equal to `INPUT_KEY`. If this syntax
```

invariant:

```tla
{{#include ../../../test/tla/bin-search/BinSearch3.tla:84:87}}
```

**Digression: Boolean connectives.** In the above code, the operator `=>` is
the [Classical implication][]. In general, `A => B` is equivalent to `IF A THEN
B ELSE TRUE`. The implication `A => B` is also equivalent to the TLA+
expression `~A \/ B`, which one can read as "not A holds, or B holds". The
operator `\/` is called *disjunction*. As a reminder, here is the standard
truth table for the Boolean connectives, which are no different from the
Boolean logic in TLA+:

| `A`     | `B`     | `~A`    | `A \/ B` | `A /\ B` | `A => B` |
| ------- | ------- | ------- | -------- | -------- | -------- |
| `FALSE` | `FALSE` | `TRUE`  | `FALSE`  | `FALSE`  | `TRUE`   |
| `FALSE` | `TRUE`  | `TRUE`  | `TRUE`   | `FALSE`  | `TRUE`   |
| `TRUE`  | `FALSE` | `FALSE` | `TRUE`   | `FALSE`  | `FALSE`  |
| `TRUE`  | `TRUE`  | `FALSE` | `TRUE`   | `TRUE`   | `TRUE`   |

**Checking Postcondition.**
Let us check `Postcondition` on `MC3_8.tla`:

```sh
$ apalache-mc check --inv=Postcondition MC3_8.tla
```

This property holds true. However, it's a small win, as `MC3_8.tla` fixes all

> parameters. Hence, we have checked the property just for one data point.  In

Reviewer 2: ```suggestion
parameters. Hence, we have checked the property just for one data point. In
```

Step 5, we will check `Postcondition` for all sequences admitted by `INT_WIDTH`.

TODO: add the link to BinSearch3.tla and MC3_8.tla when we have them on github.

## Step 4: Specifying the loop (with a caveat)

Copy `MC3_8.tla` to `MC4_8.tla`. Copy `BinSearch3.tla` to `BinSearch4.tla` and
open `BinSearch4.tla`.

We specify the loop of `binarySearch` in TLA+ as follows:

```tla

> {{#include ../../../test/tla/bin-search/BinSearch4.tla:55:78}}

Reviewer: ```suggestion
{{#include ../../../test/tla/bin-search/BinSearch4.tla:55:77}}
```

```

Let's start with these two definitions:

```tla
{{#include ../../../test/tla/bin-search/BinSearch4.tla:60:61}}
```

As you have probably guessed, we define two (local) values `mid` and `midVal`.
The value `mid` is the average of `low` and `high`. The operator `\div` is
simply integer division, which is usually written as `/` in programming

> languages. The value `midVal` is the value at the location `mid + 1`.  Since

Reviewer 2: ```suggestion
languages. The value `midVal` is the value at the location `mid + 1`. Since
```

the TLA+ sequence `INPUT_SEQ` has indices in the range `1..Len(INPUT_SEQ)`,
whereas we are computing zero-based indices, we are adjusting the index by one,
that is, we write `INPUT_SEQ[mid + 1]` instead of `INPUT_SEQ[mid]`.

*Warning: The definitions of `mid` and `midVal` do not properly reflect the
Java code of `binarySearch`. We will fix them later. It is a good exercise
to stop here and think about the source of this imprecision.*

The following lines look like ASCII graphics, but by now you should know

> enough, in order to read them:

Reviewer 2: ```suggestion
enough to read them:
```

operator `CASE` (see [Summary of TLA+][]). However, we find the disjunctive
form to be the least cluttered, though unusual.

Now we can check the postcondition again:

```sh
$ apalache-mc check --inv=Postcondition MC4_8.tla
```

The check goes through, but did it do much? Recall, that we fixed `INPUT_SEQ`
to be the empty sequence `<< >>` in `MC4_8.tla`. Hence, we never enter the loop
we have just specified.

Actually, Apalache is giving us a hint that it has never tried some of the
cases:

```
...
PASS #13: BoundedChecker
State 0: Checking 1 state invariants
Step 0: picking a transition out of 1 transition(s)
Step 1: Transition #0 is disabled
Step 1: Transition #1 is disabled
Step 1: Transition #2 is disabled
Step 1: Transition #3 is disabled
State 1: Checking 1 state invariants
Step 1: picking a transition out of 1 transition(s)
Step 2: Transition #0 is disabled
Step 2: Transition #1 is disabled
Step 2: Transition #2 is disabled
Step 2: Transition #4 is disabled
Step 2: picking a transition out of 1 transition(s)
...
```

**Digression: symbolic transitions.** Internally, Apalache decomposes the
predicates `Init` and `Next` into independent pieces like `Init == Init$0 \/

> Init$1` and `Next == Next$0 \/ Next$1 \/ Next$2 \/ Next$3`. If you want to see,
> how it is done, run Apalache with the options `--write-intermediate` and `--run-dir`:

Reviewer: ```suggestion
Init$1` and `Next == Next$0 \/ Next$1 \/ Next$2 \/ Next$3`. If you want to see
how it is done, run Apalache with the options `--write-intermediate` and `--run-dir`:
```

```sh
$ apalache-mc check --inv=Postcondition --write-intermediate=1 --run-dir=out MC4_8.tla
```

> Check the file `out/intermediate/09_OutTransition.tla`, which contains the

Reviewer: As a general question, how much can we keep printed outputs and paths in sync with our implementation?

Author: If we change the sequence of passes in the code, then the number will change.

    /\ INPUT_SEQ \in Seq(Int) 
```

This is straightforward definition. However, it does not work in Apalache:

```sh
$ apalache-mc check --cinit=ConstInit --inv=Postcondition MC5_8.tla
...
MC5_8.tla:39:22-39:29: unsupported expression: Seq(_) produces an infinite set of unbounded sequences.
Checker has found an error
...
```

The issue with our definition of `ConstInit` is that it requires the model
checker to reason about the infinite set of sequences, namely, `Seq(Int)`.
Interestingly, the model checking does not complain about the expression
`INPUT_KEY \in Int`. The reason is that this expression requires the model
checker to reason about one integer, though it ranges over the infinite set of
integers.

**Value generators.** Fortunately, this problem can be easily circumvented by
using Apalache [Value generators][], which are inspired by property-based testing,
e.g., [QuickCheck][].

Let us rewrite `ConstInit` in `MC5_8.tla` as follows:

```tla
ConstInit ==
    /\ INPUT_KEY = Gen(1)
    /\ INPUT_SEQ = Gen(MAX_INT)
```

In this new version, we use value generators to:

  - produce an unrestricted integer to be used as a value of `INPUT_KEY` and

>   - produce a  sequence of integers to be used as a value of `INPUT_SEQ`.  This

Reviewer 2: ```suggestion
  - produce a sequence of integers to be used as a value of `INPUT_SEQ`. This
```

    sequence is unrestricted, except its length is bounded with `MAX_INT`,
    which is exactly what we need in our case study.

In contrast to property-based testing, an Apalache generator is not randomly
producing values. Rather, it simply introduces a data structure whose size is
bounded with the argument of `Gen`. This lets Apalache check all instances of

> the data structure, without enumerating the instances!

Reviewer: Never having used value generators, I'm not clear on their syntax nor semantics from this explanation.

Author: Explained.

> By doing so, we are able to check the specification for all the inputs, when we
> fix the bit width. In the model `MC5_8.tla`, we fix `INT_WIDTH` to 8, to get
> the feedback fast.

Reviewer: ```suggestion
By doing so, we are able to check the specification for all the inputs, when we
fix the bit width. To quickly get feedback from Apalache, we fix `INT_WIDTH` to 8 in the model `MC5_8.tla`.
```

Let us check `Postcondition` again:

```sh
$ apalache-mc check --cinit=ConstInit --inv=Postcondition MC5_8.tla
...
State 2: state invariant 0 violated. Check the counterexample in: 
  /[a long path]/counterexample1.tla
...
```

Let us inspect the counterexample:

```tla
---------------------------- MODULE counterexample ----------------------------  
EXTENDS MC5_8

(* Constant initialization state *)
ConstInit == INPUT_KEY = -1 /\ INPUT_SEQ = <<0, -1>>

(* Initial state *)
State0 ==
  INPUT_KEY = -1
    /\ INPUT_SEQ = <<0, -1>>
    /\ high = 1
    /\ isTerminated = FALSE
    /\ low = 0
    /\ returnValue = 0

(* Transition 1 to State1 *)
State1 ==
  INPUT_KEY = -1
    /\ INPUT_SEQ = <<0, -1>>
    /\ high = -1
    /\ isTerminated = FALSE
    /\ low = 0
    /\ returnValue = 0

(* Transition 3 to State2 *)
State2 ==
  INPUT_KEY = -1
    /\ INPUT_SEQ = <<0, -1>>
    /\ high = -1
    /\ isTerminated = TRUE
    /\ low = 0
    /\ returnValue = -1
...
```

> **Fixing Postcondition.** Is it a real issue? It is, but it is not the issue of
> the search, rather our invariant `Postcondition` is imprecise.

Reviewer: ```suggestion
Is it a real issue? It is, but it is not the issue of
the search, rather our invariant `Postcondition` is imprecise.
```

Don't think this single sentence needs a section heading.

## Step 5b: Fixing the postcondition

If we check our source of truth, that is, the Java documentation in
[Arrays.java in OpenJDK][], we will see the following sentences:

    The range must be sorted (as by the {@link #sort(int[], int, int)} method)

>     prior to making this call.  If it is not sorted, the results are undefined.

Reviewer 2: ```suggestion
    prior to making this call. If it is not sorted, the results are undefined.
```

>     If the range contains multiple elements with the specified value, there is
>     no guarantee which one will be found.

Reviewer: ```suggestion
```

You're only specifying sortedness below; may want to omit this sentence.

Author: Well, this is the sentence from the Java doc. I would keep it as is.

Reviewer: You're right, sorry. I misread the English spec...

It is quite easy to add this constraint. This is where TLA+ starts to shine:

```tla

> {{#include ../../../test/tla/bin-search/BinSearch5.tla:79:88}}

Reviewer: ```suggestion
{{#include ../../../test/tla/bin-search/BinSearch5.tla:80:88}}
```

...

> {{#include ../../../test/tla/bin-search/BinSearch5.tla:109:113}}

Reviewer: ```suggestion
{{#include ../../../test/tla/bin-search/BinSearch5.tla:109:112}}
```

```

> If we check `PostconditionSorted`, we do not get any error after 10 steps:

Reviewer: Why are we checking `InputIsSorted` in the postcondition instead of constraining `Init`?

Author: This is a good question! I initially wrote the spec by constraining `INPUT_SEQ`, but then realized that we want to see the code behavior on the unsorted sequences too.

```sh
$ apalache-mc check --cinit=ConstInit --inv=PostconditionSorted MC5_8.tla
...
The outcome is: NoError
Checker reports no error up to computation length 10
```

It takes some time to explore all executions of length up to 10 steps, for all
input sequences of length up to `2^7 - 1` arbitrary integers. If we think about
it, the model checker managed to crunch infinitely many numbers in several
hours. Not bad.

*Exercise.* If you are impatient, you can check `PostconditionSorted` for the
configuration that has integer width of 4 bits. It takes only a few seconds to
explore all executions.

> TODO: add the link to BinSearch1.tla when we have it on github.

Reviewer: You probably mean a different file here

## Step 6: Checking termination and progress

Copy `BinSearch5.tla` to `BinSearch6.tla`, copy `MC5_8.tla` to `MC6_8.tla` and
open `BinSearch6.tla`.

Actually, we do not need 10 steps to check termination for the case `INT_WIDTH
= 8`. If you recall the complexity of the binary search, it needs

> `log2(Len(INPUT_SEQ))` steps to terminate.

Reviewer: You lost me here because I forgot that we let `INPUT_SEQ == Gen(MAX_INT)` and `MAX_INT == 2^(INT_WIDTH - 1) - 1`. Include a reminder?

Also, we update `Init` and `Next` in `BinSearch6.tla` as follows:

```tla
Init ==
    ...
    /\ nSteps = 0

Next ==
    IF ~isTerminated
    THEN IF low <= high
      THEN          \* lines 6-14
        /\ nSteps' = nSteps + 1
        /\ LET mid == (low + high) \div 2 IN
         ...
      ELSE          \* line 16
        /\ isTerminated' = TRUE
        /\ returnValue' = -(low + 1)
        /\ UNCHANGED <<low, high, nSteps>>
    ELSE            \* isTerminated
      UNCHANGED <<low, high, returnValue, isTerminated, nSteps>>
```

Having `nSteps`, we can write the `Termination` property:

```tla
\* We know the exact number of steps to show termination.
Termination ==
    (nSteps >= INT_WIDTH) => isTerminated
```

Let us check `Termination`:

```sh
$ apalache-mc check --cinit=ConstInit --inv=Termination MC6_8.tla
...
Checker reports no error up to computation length 10
It took me 0 days  0 hours  0 min 19 sec
```

Even if did not know precisely complexity of the binary search, we could
write a simpler property, which demonstrates progress of the search:

```tla
Progress ==
    ~isTerminated' => (low' > low \/ high' < high)
```

It takes about 10 seconds to check `Progress` as well.

> ## Step 7: Fix-width integers

Reviewer: ```suggestion
## Step 7: Fixed-width integers
```

Copy `BinSearch6.tla` to `BinSearch7.tla`, copy `MC6_8.tla` to `MC7_8.tla` and
open `BinSearch7.tla`.

Do you recall that our specification of the loop had a caveat? Let us have a
look at this piece of the specification again:

```tla
{{#include ../../../test/tla/bin-search/BinSearch6.tla:63:78}}
```

You can see that all arithmetic operations are performed over TLA+ integers,

> that is, unbounded integers. We have to implement fix-width integers ourselves.

Reviewer: ```suggestion
that is, unbounded integers. We have to implement fixed-width integers ourselves.
```

Fortunately, we do not have to implement the whole set of integer operators,
but only the addition over signed integers, which has a potential to overflow.
To this end, we have to recall how signed integers are represented in modern
computers, see [Two's complement][]. Fortunately, we do not have to worry about
an efficient implementation of integer addition. We simply use addition over

> unbounded integers to implement addition over fix-width integers:

Reviewer: ```suggestion
unbounded integers to implement addition over fixed-width integers:
```

        ...
```

This finally gives us a specification that faithfully represents the Java code
of `binarySearch`. Now we can check all expected properties once again:

```sh
$ apalache-mc check --cinit=ConstInit --inv=PostconditionSorted MC7_8.tla
...
State 2: state invariant 0 violated.
...
Total time: 2.786 sec
```

```sh
$ apalache-mc check --cinit=ConstInit --inv=Progress MC7_8.tla
...
State 1: action invariant 0 violated.
...
Total time: 2.935 sec
```

```sh
$ apalache-mc check --cinit=ConstInit --inv=Termination MC7_8.tla
...
State 8: state invariant 0 violated.
...
Total time: 39.540 sec
```

As we can see, all of our invariants are violated. They all demonstrate the
issue that is caused by the integer overflow!

TODO: add the link to BinSearchN.tla when we have it on github.

## Step 8: Checking the boundaries

Copy `BinSearch7.tla` to `BinSearch8.tla`, copy `MC7_8.tla` to `MC8_8.tla` and
open `BinSearch8.tla`.

As we have seen in Step 7, the cause of all errors in `PostconditionSorted`,
`Termination`, and `Progress` is that the addition `low + high` overflows and
thus the expression `INPUT_SEQ[mid + 1]` accesses `INPUT_SEQ` outside of its
domain.

> Why did not Apalache complain about access outside of the domain? Its behavior

Reviewer 4: ```suggestion
Why did Apalache not complain about access outside of the domain? Its behavior
```

is actually consistent with [Specifying Systems][] (p. 302):

> A function *f* has a domain DOMAIN *f*, and the value of *f*[*v*] is
> specified only if *v* is an element of DOMAIN *f*.

Hence, Apalache returns some value of a proper type, if `v` is outside of
`DOMAIN f`. As we have seen in Step 7, such a value would usually show up in a
counterexample. In the future, Apalache will probably have an automatic check
for out-of-domain access. For the moment, we can write such a check as a state
invariant. By propagating the conditions from `INPUT_SEQ[mid + 1]` up in `Next`,
we construct the following invariant:

```tla

> {{#include ../../../test/tla/bin-search/BinSearch8.tla:144:151}}

Reviewer: ```suggestion
{{#include ../../../test/tla/bin-search/BinSearch8.tla:144:150}}
```

```

> Apalache find a violation of this invariant in a few seconds:

Reviewer: ```suggestion
Apalache finds a violation of this invariant in a few seconds:
```

`Termination`, and `Progress`.

```sh
$ apalache-mc check --cinit=ConstInit --inv=InBounds MC9_8.tla
...
The outcome is: NoError
...
Total time: 76.352 sec
```

```sh
$ apalache-mc check --cinit=ConstInit --inv=Progress MC9_8.tla
...
The outcome is: NoError
...
Total time: 63.578 sec
```

```sh
$ apalache-mc check --cinit=ConstInit --inv=Termination MC9_8.tla
...
The outcome is: NoError
...
Total time: 72.682 sec
```

```sh
$ apalache-mc check --cinit=ConstInit --inv=PostconditionSorted MC9_8.tla
...
The outcome is: NoError
...
Total time: 2154.646 sec
```

*Exercise:* It takes quite a bit of time to check `PostconditionSorted`.
Change `INT_WIDTH` to 6 and check all these invariants once again. Observe that
it takes Apalache significantly less time.

*Exercise:* Change `INT_WIDTH` to 16 and check all these invariants once again.
Observe that it takes Apalache significantly more time.

TODO: add the link to BinSearchN.tla when we have it on github.

## Step 10: Beautifying the specification

Copy `BinSearch9.tla` to `BinSearch10.tla`, copy `MC9_8.tla` to `MC10_8.tla` and
open `BinSearch10.tla`.

We have reached our goals: TLA+ and Apalache helped us in finding the access

> bug and showing that its fix is working. Now it is time to look back at the

Reviewer 4: ```suggestion
bug and showing that its fix works. Now it is time to look back at the
```

specification and make it easier to read.

Let us have a look at our definition of `Next`:

```tla

> {{#include ../../../test/tla/bin-search/BinSearch9.tla:77:101}}

Reviewer: ```suggestion
{{#include ../../../test/tla/bin-search/BinSearch9.tla:77:100}}
```

```

`Next` contains a massive expression. We can decompose it nicely in smaller
pieces:

```tla

> {{#include ../../../test/tla/bin-search/BinSearch10.tla:78:114}}

Reviewer: ```suggestion
{{#include ../../../test/tla/bin-search/BinSearch10.tla:78:113}}
```

```

The definitions `LoopIteration`, `LoopExit`, and `StutterOnTermination` are
called actions in TLA+. It is usually a good idea to decompose a large `Next`
formula into actions. Normally, an action contains assignments to all primed
variables.

TODO: add the link to BinSearchN.tla when we have it on github.

## Discussion

In this tutorial we have shown how to:

 - Specify the behavior of a sequential algorithm (binary search).

>  - Specify the invariants that check safety and termination.

Reviewer: ```suggestion
 - Specify invariants that check safety and termination.
```

 - Take into account the specifics of a computer architecture (fixed bit
   width).

>  - Automatically find the examples of when all invariants are violated.

Reviewer 4: ```suggestion
 - Automatically find examples of simultaneous invariant violation.
```

>  - Efficiently check the expected properties against the fixed specification.

Reviewer: ```suggestion
 - Efficiently check the expected properties against our specification.
```

I don't understand what "the fixed" means here.

> We have written our specification for parameterized bit width. This let us
> checking the invariants relatively quickly and get fast feedback from the model

Reviewer 4: ```suggestion
We have written our specification for parameterized bit width. This let us
check the invariants relatively quickly and get fast feedback from the model
```

> checker. We chose the bit width of 8, as it is non-trivial, while Apalache does
> not get stuck while running for this bit width. Importantly, the specification

Reviewer 4: ```suggestion
checker. We chose the bit width of 8, as it is a non-trivial value, for which Apalache terminates in reasonable time. Importantly, the specification
```

Reviewer: Or rather

```suggestion
checker. We chose a bit width of 8, a non-trivial value for which
Apalache terminates within reasonable time. Importantly, the specification
```

> for the bit width of 32 stays the same; we only have to change `INT_WIDTH`. Of
> course, Apalache reaches its limits, when we set `INT_WIDTH` to 16 or 32. In
> these cases, it has to reason about all sequences of length up to 32,767
> elements and 2 Billion elements respectively!

Reviewer: ```suggestion
for the bit width of 32 stays the same; we only have to change `INT_WIDTH`. Of
course, Apalache reaches its limits when we set `INT_WIDTH` to 16 or 32. In
these cases, it has to reason about all sequences of length up to 32,767
elements or 2 Billion elements, respectively!
```

Although Apalache gives us a good idea about whether the properties of the
binary search hold true, it does give us an ultimate proof of correctness for

> Java integers. If you need such a proof, you should probably use TLAPS.  Check

Reviewer 2: ```suggestion
Java integers. If you need such a proof, you should probably use TLAPS. Check
```

the paper on [Proving Safety Properties][] by Leslie Lamport.

This tutorial is rather long. This is because we wanted to show the evolution
of a TLA+ specification, as we were writing it and checking it with Apalache.
It took us 2-3 hours to iteratively develop a specification that is similar to
the one demonstrated in this tutorial.

> Most likely, there are many different styles of writing TLA+ specifications.

Reviewer 4: ```suggestion
There are many different styles of writing TLA+ specifications.
```

One of our goals was to demonstrate the incremental approach to specification
writing. In fact, this approach is not very different from incremental
development of programs in the spirit of [Test-driven development][].

This tutorial touches upon the basics of TLA+ and Apalache. For instance, we
did not discuss non-determinism, as our specification is entirely

> deterministic. We will demonstrate advanced features in the future tutorials.

Reviewer 4: ```suggestion
deterministic. We will demonstrate advanced features in future tutorials.
```
